File: bindings/lua/tests/smoke.lua
-- SPDX-License-Identifier: Apache-2.0
-- Copyright (c) 2024-2026 Quad4.io

local rns = require("rns")

local function assert_eq(a, b, msg)
	if a ~= b then
		error((msg or "assert_eq") .. ": got " .. tostring(a) .. " want " .. tostring(b), 2)
	end
end

local function assert_true(cond, msg)
	if not cond then
		error(msg or "assert_true failed", 2)
	end
end

assert_eq(rns.version(), rns.API_VERSION, "version")

do
	local node = rns.Node.create()
	local ok, err = pcall(function()
		node:start()
		local ok2, err2 = pcall(function()
			node:event_poll(10)
		end)
		assert_true(not ok2, "expected timeout")
		assert_eq(err2.code, rns.Error.TIMEOUT, "timeout code")
		node:stop()
	end)
	node:close()
	assert_true(ok, err and (err.message or tostring(err)) or "node lifecycle")
end

do
	local identity = rns.Identity.generate()
	local hex_hash = identity:hash_hex()
	assert_eq(#hex_hash, 32, "hash hex length")
	assert_eq(#identity:hash_bytes(), 16, "hash bytes length")
	local msg = "hello-rns"
	local sig = identity:sign(msg)
	identity:verify(msg, sig)
	local pub = identity:public_key()
	local only_pub = rns.Identity.from_public_key(pub)
	only_pub:verify(msg, sig)
	only_pub:close()
	identity:close()
end

print("lua smoke tests ok")

File: bindings/rust/tests/smoke.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2024-2026 Quad4.io

use rns::{version, Identity, Node, API_VERSION, Error};

#[test]
fn version_matches_api() {
    assert_eq!(version(), API_VERSION);
}

#[test]
fn node_lifecycle() {
    let node = Node::create("").expect("node create");
    node.start().expect("start");
    assert!(matches!(node.event_poll(10), Err(Error::Timeout)));
    node.stop().expect("stop");
}

#[test]
fn identity_generate_hash_sign() {
    let id = Identity::generate().expect("generate");
    let hex = id.hash_hex().expect("hash");
    assert_eq!(hex.len(), 32);
    let bytes = id.hash_bytes().expect("hash bytes");
    assert_eq!(bytes.len(), 16);
    let msg = b"hello-rns";
    let sig = id.sign(msg).expect("sign");
    id.verify(msg, &sig).expect("verify");
    let pub_key = id.public_key().expect("pubkey");
    let only_pub = Identity::from_public_key(&pub_key).expect("from pubkey");
    only_pub.verify(msg, &sig).expect("verify via pubkey");
}
